include/inviwo/core/datastructures/tfprimitiveset.h

```diff
@@ -0,0 +1,182 @@
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#ifndef IVW_TFPRIMITIVESET_H
+#define IVW_TFPRIMITIVESET_H
+
+#include <inviwo/core/common/inviwocoredefine.h>
+#include <inviwo/core/common/inviwo.h>
+
+#include <inviwo/core/datastructures/tfprimitive.h>
+#include <inviwo/core/util/observer.h>
+
+namespace inviwo {
+
+enum class TFPrimitiveSetType {
+    Relative,  //<! uses the normalized range [0,1] for all TF primitives
+    Absolute,  //<! absolute positioning of TF primitives
+};
+
+class IVW_CORE_API TFPrimitiveSetObserver : public Observer {
+public:
+    virtual void onTFPrimitiveAdded(TFPrimitive* p);
+    virtual void onTFPrimitiveRemoved(TFPrimitive* p);
+    virtual void onTFPrimitiveChanged(const TFPrimitive* p);
+};
+class IVW_CORE_API TFPrimitiveSetObservable : public Observable<TFPrimitiveSetObserver> {
+protected:
+    void notifyTFPrimitiveAdded(TFPrimitive* p);
+    void notifyTFPrimitiveRemoved(TFPrimitive* p);
+    void notifyTFPrimitiveChanged(const TFPrimitive* p);
+};
+
+/**
+ * \ingroup datastructures
+ * \class TFPrimitiveSet
+ * \brief data structure managing multiple TFPrimitives
+ *
+ * \see TFPrimitive
+ */
+class IVW_CORE_API TFPrimitiveSet : public Serializable,
+                                    public TFPrimitiveSetObservable,
+                                    public TFPrimitiveObserver {
+public:
+    TFPrimitiveSet(const std::vector<TFPrimitiveData>& values = {},
+                   TFPrimitiveSetType type = TFPrimitiveSetType::Relative);
+    TFPrimitiveSet(const TFPrimitiveSet& rhs);
+    TFPrimitiveSet(TFPrimitiveSet&& rhs);
+    TFPrimitiveSet& operator=(const TFPrimitiveSet& rhs);
+    virtual ~TFPrimitiveSet() = default;
+
+    TFPrimitiveSetType getType() const;
+
+    size_t size() const;
+
+    TFPrimitive* operator[](size_t i);
+    const TFPrimitive* operator[](size_t i) const;
+
+    TFPrimitive* get(size_t i);
+    const TFPrimitive* get(size_t i) const;
```

Comment:
Would be nice with begin() and end() iters...

Reply:
hm, would that then be on the unsorted or sorted primitives?

Reply:
choosing sorted for now.